File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool is_palindrome(string mystr){
        int start = 0, end = mystr.size()-1;
        while (start < end) {
            if (mystr[start] != mystr[end]) return false;
            start++;
            end--;
        }
        return true;
    }
    
    void backtrack(string s, int start, vector<string> path, vector<vector<string>>& result ){
        if (start == s.size()){
            result.push_back(path);
            return;
        }
        for (int idx=start; idx < s.size(); idx++){
            int end = idx;
            string mysub = s.substr(start, end-start+1);
            if (is_palindrome(mysub)){
                path.push_back(mysub);
                backtrack(s, end+1, path, result);
                path.pop_back();
            }
        }
        
    }
    vector<vector<string>> partition(string s) {
        vector<string> path;
        vector<vector<string>> result;
        backtrack(s, 0, path, result);
        return result;        
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """

        def is_palindrome(mystr):
            start, end = 0, len(mystr)-1
            while start < end :
                if mystr[start] != mystr[end] :
                    return False
                start+=1
                end-=1
            return True


        def backtrack_pal(s, start, path, result):
            if start == len(s):  
                result.append(path[:])
                return
            
            for end in range(start, len(s)):
                if is_palindrome(s[start:end+1]):
                    path.append(s[start:end+1]) 
                    backtrack_pal(s, end+1, path, result)  
                    path.pop() 

        result = []
        backtrack_pal(s, 0, [], result)
        return result